components/encryption/src/master_key/kms.rs

```diff
@@ -0,0 +1,306 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::fs::create_dir_all;
+use std::future::Future;
+use std::path::Path;
+use std::time::Duration;
+
+use futures::future::{self, TryFutureExt};
+use kvproto::encryptionpb::{EncryptedContent, EncryptionMethod};
+use rusoto_core::region;
+use rusoto_core::request::DispatchSignedRequest;
+use rusoto_core::request::HttpClient;
+use rusoto_credential::{DefaultCredentialsProvider, StaticProvider};
+use rusoto_kms::{DecryptRequest, GenerateDataKeyRequest, Kms, KmsClient};
+use tokio::runtime::{Builder, Runtime};
+
+use super::{metadata::MetadataKey, Backend, FileBackend, PlainTextBackend, WithMetadata};
+use crate::config::KmsConfig;
+use crate::encrypted_file::EncryptedFile;
+use crate::{Error, Result};
+
+const KMS_ENCRYPTION_KEY_NAME: &str = "kms_encryption.key";
+
+// Always use AES 256 for encrypting master key.
+const KMS_DATA_KEY_METHOD: EncryptionMethod = EncryptionMethod::Aes256Ctr;
+const AWS_KMS_DATA_KEY_SPEC: &str = "AES_256";
+const AWS_KMS_VENDOR_NAME: &[u8] = b"AWS";
+
+struct AwsKms {
+    client: KmsClient,
+    current_key_id: String,
+}
+
+impl AwsKms {
+    fn new(config: KmsConfig) -> Result<AwsKms> {
+        let http_dispatcher = HttpClient::new().unwrap();
+
+        AwsKms::with_request_dispatcher(config, http_dispatcher)
+    }
+
+    // TODO following code is almost the same as external_storage s3 client,
+    //      should be wrapped together.
+    fn with_request_dispatcher<D>(config: KmsConfig, dispatcher: D) -> Result<AwsKms>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        if config.key_id.is_empty() {
+            return Err(Error::Other(
+                "KMS key id can not be empty".to_owned().into(),
+            ));
+        }
+
+        let region = if config.endpoint.is_empty() {
+            config.region.parse::<region::Region>().map_err(|e| {
+                Error::Other(format!("invalid region format {}: {}", config.region, e).into())
+            })?
+        } else {
+            region::Region::Custom {
+                name: config.region,
+                endpoint: config.endpoint,
+            }
+        };
+
+        let client = if config.access_key.is_empty() || config.secret_access_key.is_empty() {
+            let cred_provider = DefaultCredentialsProvider::new()
```

Comment:
Please update to use the interface in #7201 after that PR is merged.

Reply:
Sure